metrics/framework/Metrics.cpp

```diff
@@ -0,0 +1,212 @@
+/*##############################################################################
+    HPCC SYSTEMS software Copyright (C) 2021 HPCC Systems®.
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+       http://www.apache.org/licenses/LICENSE-2.0
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "Metrics.hpp"
+
+using namespace hpccMetrics;
+
+void CounterMetric::getReportValues(MeasurementVector &mv) const
+{
+    mv.emplace_back((std::make_shared<Measurement<uint32_t>>(name, description, ValueType::METRICS_INTEGER, count.load())));
+}
+
+
+void GaugeMetric::getReportValues(MeasurementVector &mv) const
+{
+    mv.emplace_back((std::make_shared<Measurement<float>>(name, description, ValueType::METRICS_FLOAT, gaugeValue.load())));
+}
+
+
+IMetricSink *MetricSink::getSinkFromLib(const char *type, const char *sinkName, const IPropertyTree *pSettingsTree)
+{
+    return MetricSink::getSinkFromLib(type, "", sinkName, pSettingsTree);
+}
+
+IMetricSink *MetricSink::getSinkFromLib(const char *type, const char *getInstanceProcName, const char *sinkName, const IPropertyTree *pSettingsTree)
+{
+    std::string libName;
+
+    //
+    // First, treat type as a full library name
+    libName = type;
+    if (libName.find(SharedObjectExtension) == std::string::npos)
+    {
+        libName.append(SharedObjectExtension);
+    }
+    HINSTANCE libHandle = LoadSharedObject(libName.c_str(), true, false);
+
+    //
+    // If type wasn't the lib name, treat it as a type
```

Comment:
why?  We can define what the name means.  What is the situation where this is needed?

Reply:
This was added during review with Rigo. The thought was that you may package a number of different features into a single lib and this gives a way to extract the sink functionality.

Reply:
I'm not sure if this is one of the suggestions I provided; but yes I pushed a few enhancements related to plugin library nomenclature flexibility which resolved issues often encountered in practice (mostly in the pluggable security framework). If the criteria is to trim out any/all unnecessary logic, we can certainly employ a strict protocol regarding library names, etc.

Reply:
If there is a clear description of the problem it is solving, and that problem seems reasonable, then it is fine.  We load plugins in lots of other places and don't have alternatives.
Since type is going in a helm chart it should be something descriptive, and probably some kind of calculated filename.